Editorial: Add Bikeshed metadata

mikewest opened the pull request:
This patch adds Bikeshed metadata for various concepts and terms
used by CSP and related specifications.

mikewest commented:
This got a bit large, but I think it covers everything CSP needs (along with a few others that I exported while I was there).

annevk commented on source:
```diff
@@ -2167,12 +2167,12 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
-  initialised to true. <ref spec=DOM></p>
+  the event through the tree. The term <dfn data-x="concept-events-trusted" data-export="">trusted
+  event</dfn> is used to refer to events whose <code data-x="dom-event-isTrusted">isTrusted</code>
+  attribute is initialised to true. <ref spec=DOM></p>
```
I don't think we should export these. Folks should use the definitions from the DOM Standard directly.

mikewest replied:
Ok. Looking at CSP I can probably simplify the callsite anyway.

annevk commented on source:
```diff
@@ -8715,7 +8716,7 @@ http://software.hixie.ch/utilities/js/live-dom-viewer/?%3C%21DOCTYPE%20html%3E..
 
 [OverrideBuiltins]
-partial /*sealed*/ interface <dfn>Document</dfn> {
+partial /*sealed*/ interface <dfn data-dfn-type=interface data-export="">Document</dfn> {
```
This would conflict with DOM, no? This is only a partial interface after all.

mikewest replied:
Hrm. You're right. I guess we can just point to the main interface, and to the individual attributes here if we ever need them. Will change.

tabatkins replied:
Yes, exactly. Only the main defining interface should dfn itself; partials link.  Only the things actually getting defined get dfn'd.

mikewest replied:
@annevk: Should I drop the existing `<dfn>` in this patch, or is that something that HTML's build system relies on in some way?

annevk commented on source:
```diff
@@ -12103,7 +12104,8 @@ gave me some of the songs they wrote. I love sharing my music.&lt;/p>
 
-  <h4 id="the-head-element">The <dfn><code>head</code></dfn> element</h4>
+  <h4 id="the-head-element">The <dfn data-export="" data-dfn-type=element
+  id=elementdef-head><code>head</code></dfn> element</h4>
```
Thus far we quote all attribute values. Not sure if we want to deviate from that. @domenic?

annevk replied:
Also, you should add the attributes to the `<h4>` element (here and below). That is the one we actually use and want folks to reference.

mikewest replied:
> Thus far we quote all attribute values.

*sobs* How do y'all possibly work on HTML and other specs at the same time? The style divergences must drive you nuts. :)

annevk replied:
I agree that we should probably try to decide on something here, especially given the increase in outside contributors. Not really sure how to best raise that.

mikewest replied:
> Also, you should add the attributes to the <h4> element (here and below). That is the one we actually use and want folks to reference.

Hrm. This seems inconsistent in the current text, and also seems like it's going to cause some confusion. Look at `script`, for instance. That's currently marked up with a `<dfn id="script">`, and would have a linking text of "script". If we changed the markup to somethign like `<h4 id="the-script-element" data-dfn-type="element">The <dfn id="script">script</dfn> element</h4>`, then Bikeshed would be looking for something like `<{the script element}>`. If we add `data-lt="script"`, then Bikeshed would likely complain about two definitions for the same term.

Marking up the `<dfn>` seems cleaner to me, but I'm happy to run with whatever y'all like.

annevk replied:
I'm not sure why the `script` element has the additional ID there. All references go to `#the-script-element`. Might be a legacy ID. Since the inner one is not exported, I believe Bikeshed would not complain.

domenic replied:
Yeah everything goes to `#the-script-element` or similar for every other element; it indeed seems like a legacy ID. The other `<dfn>`s inside `<h4>`s that I checked do not have IDs.

mikewest replied:
Done in latest patch (`<meta>` has an ID too, FWIW).

annevk commented on source:
```diff
@@ -112068,7 +112088,8 @@ if (s = prompt('What is your name?')) {
 
-  <h4>The <dfn><code>applet</code></dfn> element</h4>
+  <h4>The <dfn data-export="" data-dfn-type="element" id=elementdef-applet><code>applet</code></dfn>
+  element</h4>
```
This is still wrong.

annevk replied:
As in, it should be on `<h4>`.

annevk replied:
And no new ID.

mikewest replied:
Fixed in latest, thanks!

annevk commented on source:
```diff
@@ -112384,8 +112405,9 @@ if (s = prompt('What is your name?')) {
-  <p>The <dfn><code>frameset</code></dfn> element acts as <span>the body element</span> in documents
-  that use frames.</p>
+  <p>The <dfn data-export="" data-dfn-type="element"
+  id="elementdef-frameset"><code>frameset</code></dfn> element acts as <span>the body element</span>
```
This changes its current ID away from "frameset".

mikewest replied:
Fixed in latest.

annevk commented on source:
```diff
@@ -112412,9 +112434,9 @@ if (s = prompt('What is your name?')) {
-  <p>The <dfn><code>frame</code></dfn> element defines a <span>nested browsing context</span>
-  similar to the <code>iframe</code> element, but rendered within a <code>frameset</code>
-  element.</p>
+  <p>The <dfn data-export="" data-dfn-type="element" id="elementdef-frame"><code>frame</code></dfn>
```
Its current ID is "frame".

mikewest replied:
Fixed in latest.

annevk commented on source:
```diff
@@ -12877,7 +12878,7 @@ interface <dfn>HTMLLinkElement</dfn> : <span>HTMLElement</span> {
 
 
-  <h4 id="the-meta-element">The <dfn id="meta"><code>meta</code></dfn> element</h4>
+  <h4 data-dfn-type="element" data-lt="meta">The <dfn id="meta"><code>meta</code></dfn> element</h4>
```
Why did you remove the id from the `<h4>`?

mikewest replied:
Because it gets autogenerated from the element's textual content? Maybe I'm misunderstanding, but the current text in the spec does not have an ID for every `<h4>` element. When are you ok with autogenerated IDs, and when should we hard-code them?

annevk replied:
If it's there we shouldn't mess with it, basically.

mikewest replied:
If it produces the same output, doesn't it seem better to be consistent in the document? We should either add IDs manually or not, but doing both at random is hard to understand.

*shrug* If you want me to put them back, I will. It just doesn't make sense to me. :)

annevk replied:
Are you sure it produces the same output? I'm pretty sure removing the one for "the head element" actually changed the output.

mikewest replied:
Reinstated in latest patch.

annevk commented on source:
```diff
@@ -12103,7 +12104,7 @@ gave me some of the songs they wrote. I love sharing my music.&lt;/p>
   <h3>Document metadata</h3>
 
-  <h4 id="the-head-element">The <dfn><code>head</code></dfn> element</h4>
+  <h4 data-dfn-type="element" data-lt="head">The <dfn><code>head</code></dfn> element</h4>
```
You should not remove the id from the h4.

mikewest replied:
Ditto.

annevk commented:
There's no need for export on `<h4>`?

mikewest commented:
> There's no need for export on `<h4>`?

According to https://github.com/tabatkins/bikeshed/blob/master/docs/dfn-contract.md, "If neither is provided, `dfn` type definitions default to noexport, while all others default to export".

domenic commented:
But these are "element" type definitions, not "dfn" type, right?

mikewest commented:
Right. My understanding is that they will default to export, so we don't need to add `data-export=""` to the `h4`s. I guess we don't need to add it to the element attributes either, now that I think about it.

mikewest commented:
Hrm. After that last patch, the preprocessor is failing with errors I don't understand. :/

mikewest commented:
Ah. The preprocessor is looking for `<code>([^<]+)</code></dfn> elements?</h4>`, so the line break breaks its detection (in `.pre-process-tag-omission.pl`, line 30). That seems brittle. :(

mikewest commented:
I think the latest patch addresses the remaining concerns? I put the IDs back, and then pulled them back onto one line so as not to break the pre-processor. Is there anything else you'd like to see change?
